### PR TITLE
[SYCL] Add arithmetic overloaded operators for annotated_arg

### PR DESCRIPTION
**Pull request by yug-intel**
Adding arithmetic overloaded operators for annotated_arg.

The motivation for adding these is to improve user experience when templating `annotated_arg` with other templated classes that support arithmetic operators. Example, without this change. the following SYCL code results in errors:
```c++
// Some class that has arithmetic operators as member functions.
template<int W, bool S>
class ac_int;

struct MyKernel {
annotated_arg<ac_int<5, true>, ...> arg_a;
annotated_arg<ac_int<5, true>, ...> arg_b;
annotated_arg<ac_int<5, true>*, ...> arg_c;
void operator()() const {
  *arg_c = arg_a + arg_b; // error!
}
};
```
After this change, the above code will not result in errors as the new operators will kick in and invoke the add operator from `ac_int`.
Alternatively, we considered what if we direct users to create free floating operators for such cases. Eg, if we had:
```c++
template<...>
...   operator+(ac_int<...> a, ac_int<...> b) { ... }
```
This style still doesn't despite `annotated_arg` having an implicit conversion operator (due to some C++ rules about implicit conversion I think):
https://godbolt.org/z/h5cTTr17K

We are adding these generic operators here so that users do not have to create these operators themselves.

### CODE REVIEW

**Comment by sergey-semenov** [comment]
Does the draft spec document need to be updated to include these?

**Comment by yug-intel** [comment]
@sergey-semenov I updated the spec doc. Let me know if the change is good.

**Comment by tiwaria1** [comment]
The motivation for adding these is to improve user experience when templating `annotated_arg` with other templated classes that support arithmetic operators. Example, without this change. the following SYCL code results in errors:
```c++
// Some class that has arithmetic operators as member functions.
template<int W, bool S>
class ac_int;

struct MyKernel {
annotated_arg<ac_int<5, true>, ...> arg_a;
annotated_arg<ac_int<5, true>, ...> arg_b;
annotated_arg<ac_int<5, true>*, ...> arg_c;
void operator()() const {
  *arg_c = arg_a + arg_b; // error!
}
};
```
After this change, the above code will not result in errors as the new operators will kick in and invoke the add operator from `ac_int`.
Alternatively, we considered what if we direct users to create free floating operators for such cases. Eg, if we had:
```c++
template<...>
...   operator+(ac_int<...> a, ac_int<...> b) { ... }
```
This style still doesn't despite `annotated_arg` having an implicit conversion operator (due to some C++ rules about implicit conversion I think):
https://godbolt.org/z/h5cTTr17K

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (tiwaria1) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:369]: You will also have to add these operators to the section above that has the class skeleton.

File: sycl/include/sycl/ext/oneapi/experimental/annotated_arg/annotated_arg.hpp
Comment (tiwaria1) [inline sycl/include/sycl/ext/oneapi/experimental/annotated_arg/annotated_arg.hpp:166]: I initially thought this is the same propertyList as the one that templates this class. On closer look turns out that's not the case. Still rename it for clarity. We need these to work across different property lists, for example:
```c++
struct MyKernel {
  // NOTE: all arguments have different properties.
  annotated_arg<ac_int<...>, decltype(properties(A))> arg_a;
  annotated_arg<ac_int<...>, decltype(properties(B))> arg_b;
  annotated_arg<ac_int<...>, decltype(properties(C))> arg_c;
  void operator()() const {
    *arg_c = arg_a + arg_b;
  }
};
```
The properties for this class are only meaningful on the interface so it should be legal to allow arithmetic ops across objects with different property lists.

```suggestion
  template <typename T2, typename PropertyList2>
```

And similarly below and update the extension too.

**Review by tiwaria1 — COMMENTED** [review]
Seems good to me. Wait for review from Joe too. Will discuss testing offline with you.

**Comment by broxigarchen** [comment]
Do you need to add all binary operators? including <<, >>,|| and etc

**Comment by broxigarchen** [comment]
please update the e2e test here https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/Annotated_arg_ptr/annotated_arg.cpp

**Review by broxigarchen — CHANGES_REQUESTED** [review]
please update the e2e test here https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/Annotated_arg_ptr/annotated_arg.cpp

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:374]: Several questions:

* What is the return type for these operators?  Just saying `auto` isn't very helpful.  In particular, is the return type `T` or `annotated_ptr<T>`?  If the latter, what are the properties?
* These operators all assume the RHS is `annotated_arg`.  What if the user wants to do `operator+` where the RHS is a raw type without `annotated_arg`?
* What are the constraints on the type `T2`?
* Can't we just rely on the implicit conversion from `annotated_arg<T>` to `T`, rather than adding all these operators?

Comment (tiwaria1) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:374]: > Can't we just rely on the implicit conversion from annotated_arg<T> to T, rather than adding all these operators?

Greg, see my comment on the PR (also added to description) regarding this one.

Comment (tiwaria1) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc]: > What is the return type for these operators? Just saying auto isn't very helpful.

I think the return type necessarily has to be auto if we want to keep this generic. For example take the case where `annotated_arg` is templated on Intel's `ac_int` library then multiplication returns a wider type the size of which is decided by the `ac_int::operator*`.

> In particular, is the return type T or annotated_ptr<T>? If the latter, what are the properties?

Returned type would not be an `annotated_arg`

Comment (gmlueck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc]: I think it would help to specify the constraints and the return type better.  For example:

> Available only if `operator+(T2)` is valid for objects of type `T`.  The value and type of the result is the same as the result of `operator+(T2)` applied to the object of type `T`.

Comment (GarveyJoe) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc]: > I tested this and it would also require additional overloaded operators with a generic type such as this:

Use SFINAE:
```
template<typename T, typename T2, typename R = decltype(std::declval<T>() + std::declval<T2>())>
R operator+(annotated_arg<T> a, T2 b) {
    return a + b;
};
```
I think this will allow you to use an object of type annotated_arg<T> in any operator+ in which you could have used an object of type T.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (GarveyJoe) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:229]: This isn't useful when I want to add an annotated_arg<T> to a T.  See my comment in Greg's chain for a solution to that.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (yug-intel) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc]: @GarveyJoe I tested this notation and the following code works: 
```cpp
template<typename T, typename PropertyList, typename T2, typename R = decltype(std::declval<T>() + std::declval<T2>())>
R operator+(const annotated_arg<T, PropertyList> &a, T2 b) {
    T a1 = a;
    return a1 + b;
};
```

Can you confirm that we want to add these operators in addition to the operators I've currently added for `annotated_arg<T1> + annotated_arg<T2>`?

Comment (GarveyJoe) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc]: Yes, confirmed

File: sycl/test-e2e/Annotated_arg_ptr/annotated_arg.cpp
Comment (tiwaria1) [inline sycl/test-e2e/Annotated_arg_ptr/annotated_arg.cpp:34]: Testing with ints is not sufficient. Create some templated struct like:
```c++
template<typename T>
struct MyStruct{
  T data;
  // ... arithmetic operators
}
```
Add tests to check these operations for non pointer types. I think header changes might be needed if you have only added the operators for the `annotated_arg<T*>` specialization. Specifically check these:
```c++

annotated_arg<MyStruct, ...> a;
annotated_arg<MyStruct, ...> b;
MyStruct c;
q.single_task([=]{
  ... = a + b;
  ... = b + c;

}).wait();
```

**Review by tiwaria1 — CHANGES_REQUESTED** [review]
Commented.

**Comment by yug-intel** [comment]
@tiwaria1 @GarveyJoe The compound overloaded operators are not able to be used in the kernel code as they are marked const. For example:

```c++
int main() {
  queue q;

  annotated_arg<int> a = 5;
  annotated_arg<int> b = 2;

  q.single_task([=]() {
    a += b;
  }).wait();

  assert(a == 7);

  return 0;
}
```
This code produces the following error and note: 
```
temp.cpp:14:5: error: no viable overloaded '+='
   14 |                 a += b;
      |                 ~ ^  ~
/p/psg/swip/w/yugteshs/sycl_include/sycl/ext/oneapi/annotated_arg/annotated_arg.hpp:307:8: note: candidate function template not viable: 'this' argument has type 'const annotated_arg<int>', but method is not marked const
  307 |   void operator+=(const annotated_arg<T2, PropertyList2> &other) {
 ```

The overloaded operator+= works outside of the kernel code. If these operators are not usable within the kernel code, should we remove them?

**Comment by yug-intel** [comment]
Sorry closed accidentally

**Comment by GarveyJoe** [comment]
> The overloaded operator+= works outside of the kernel code. If these operators are not usable within the kernel code, should we remove them?

I don't think there's any value to overloading the += operator, or any of these computation + assignment type operators, for annotated_arg.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (yug-intel) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc]: I tested this case and it seems like the `annotated_arg<T>` is implicitly converted to `T2` and then the overloaded operator for `T2` is used. So there is no need to add an overloaded operator for this case.

Comment (gmlueck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc]: It seems very surprising to me that you can rely on implicit conversion for:

```
ac_int<int> ac1{1}, ac2{2};
annotated_arg<ac_int<int>> ann_a1{ac1}, ann_a2{ac2};
ac_int<int> ac3 = ac1 + ann_a2;
```

while at the same time implicit conversion doesn't work for:

```
ac_int<int> ac1{1}, ac2{2};
annotated_arg<ac_int<int>> ann_a1{ac1}, ann_a2{ac2};
ac_int<int> ac3 = ann_a1 + ac2;
```

Are you sure about this?  If so, what causes the assymetry?

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (gmlueck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:537]: Yes, much better!

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (yug-intel) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc]: I tested this and the assymetry is present anytime a class is templated on a user-defined class. For this code: 
```c++
struct Custom {
    int value;
    Custom(int value) : value(value){};
    Custom operator+(const Custom &rhs) const {
        return Custom(value + rhs.value);
    }
};

template<typename T>
class CustomClass {
    T value;
    CustomClass(T val) : value(val) {}
    operator T() {
        return value;
    }
};

int main() {
    CustomClass<Custom> myInt(Custom(42));
    Custom result = myInt + Custom(8);
    assert(result.value == 50 && "result value does not match");
    return 0;
}
```
This produces the following error: 
```
temp.cpp:31:27: error: invalid operands to binary expression ('CustomClass<Custom>' and 'Custom')
   31 |     Custom result = myInt + Custom(8);
      |                     ~~~~~ ^ ~~~~~~~~~
```
If the line is instead: `Custom result = Custom(8) + myInt`, there is no error. Also, if the templated parameter for CustomClass is int rather a user-defined class, the above error is not produced. I am not sure what is the cause for this assymetry when a user-defined class is used as the template parameter.

Comment (gmlueck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc]: Do you want to handle a case like this?

```
#include <utility>

template<typename T>
struct Custom {
    T value;
    Custom(T value) : value(value){};
};

template<typename T>
Custom<T> operator+(const Custom<T> &lhs, const Custom<T> &rhs) {
    return Custom<T>(lhs.value + rhs.value);
}

template<typename T>
struct annotated_arg {
    T value;
    annotated_arg(T val) : value(val) {}
    operator T() {
        return value;
    }
};

template<typename T, typename T2,
         typename R = decltype(std::declval<T>() + std::declval<T2>())>
R operator+(const annotated_arg<T> &a, const T2 &b) {
    return a.value + b;
}

#if 0
template<typename T, typename T2,
         typename R = decltype(std::declval<T>() + std::declval<T2>())>
R operator+(const T &a, const annotated_arg<T2> &b) {
    return a + b.value;
}
#endif

int main() {
    annotated_arg<Custom<int>> annotatedCustom(Custom(42));
    Custom<int> r1 = annotatedCustom + Custom(8);
    Custom<int> r2 = Custom(8) + annotatedCustom;
    return 0;
}
```

(https://godbolt.org/z/TYsfEsEh1)

Comment (yug-intel) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc]: It is strange how this only causes an error when the overloaded operator for `Custom<T>` is a free floating function. I tested this with emulator compilation of the `annotated_arg.cpp` file but I did not see the same issue. I changed the overloaded operators of `MyStruct<T>` to be free floating and tested `MyStruct<T> + annotated_arg<MyStruct<T>>` and it did not produce the invalid operands error. It seems that this problem does not occur with the `annotated_arg` but we may want to add overloaded operators for raw type on lhs to be safe.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (GarveyJoe) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:255]: Is there any particular reason why you declared these as free functions but the annotated_arg op annotated_arg versions as member functions?  As far as I can tell the versions that start with annotated_arg won't behave differently regardless of where you declare them because the difference between the two versions is whether implicit conversions are considered but implicit conversions aren't considered anyways for function templates.  Is there something I'm missing or is this decision truly arbitrary?

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (GarveyJoe) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:445]:
```suggestion
applied to the underlying objects of type `T` and `T2`.
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (yug-intel) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:262]: There does not seem to be any functional differences in this case but it is generally recommended to use non-member function when possible as they offer looser coupling. Additionally, since we the overloaded operators for when the raw type is on the lhs must be non-member functions, using non-member functions for all overloaded operators of annotated_arg and raw type will be more consistent.

**Review by tiwaria1 — APPROVED** [review]
LGTM, requested Yug to add logical operator overloads too but I do not mind if that is a separate PR.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (yug-intel) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:514]: @gmlueck I have added some more overloaded operators for logical and bit shift operations. I thought that the spec doc was getting very repetitive so I described the overloaded operators in a generic way before the table. Although, I am wondering whether the table itself is repetitive. We have listed all the overloaded operators in the class definition above so maybe here we can just have the description or have a generic template such as this: 

```c++
template <typename T2, typename PropertyList2,
          typename R = decltype(std::declval<T>() OP std::declval<T2>())>
R operatorOP(const annotated_arg<T2, PropertyList2> &other) const;
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (yug-intel) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:617]: @gmlueck Also here, I have made the table such that the operator definition with annotated_arg on LHS is on one column and the definition with annotated_arg on RHS is on the other column. Let me know if this looks ok.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (GarveyJoe) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:514]: I like this idea and think it would make the spec more readable.

**Review by ajaykumarkannan — APPROVED** [review]

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (GarveyJoe) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:514]: I think the use of the word "template" is confusing here given that it seems you're using it for its English definition when the term "template" has a meaning in C++ and the set of functions you're describing are themselves function templates.

Perhaps just add this as a single row in the table (then it's clear that it refers to a member function) but put operatorOP in italics and in the text in the right side column say something to the effect of "operatorOP can be any of ..." and list all the ops.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (GarveyJoe) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:518]:
```suggestion
The value and result is the same as the result of `operatorOP(T2)` applied to the
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (GarveyJoe) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:528]: For the same reason as above, this is a little confusing to read.  It sounds like you're saying that this is the code itself that will be in the header file as opposed to a collapsed representation of it.  Maybe go with:

```suggestion
Free functions of the following form are defined for the `overloaded operators for types `annotated_arg<T, ...>`
```

And, as above, I think you should work in the list of functions somewhere.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (GarveyJoe) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:532]:
```suggestion
of `operatorOP(T, T2)` applied to the underlying objects of type `T` and `T2`.
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (GarveyJoe) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:526]: This still isn't in the table.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (GarveyJoe) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:514]: If you aren't going to list the overloaded operators I think you at least need to say:
```suggestion
Free functions of the following form are defined for the overloaded operators listed above for
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (yug-intel) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc]: Added the table.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc
Comment (GarveyJoe) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_arg.asciidoc:525]: By my reading of this text, only member functions of T can be called, but that's not what the template checks for and it's not how you implemented it in the header files.  The text implies it was written like this:

```
  template <typename T2, typename PropertyList2,
    typename R = decltype(std::declval<T>() + std::declval<T2>())>
  R operator+(const annotated_arg<T2, PropertyList2> &other) const {
    return obj.operatorOP(other.obj);
  }
```

I think if you want it to match what you implemented you should instead say:

```suggestion
The overloaded operator `operatorOP` utilizes `operatorOP(T, T2)` and is available only if `operatorOP(T, T2)` is well formed.  
The value and result is the same as the result of `operatorOP(T, T2)` applied to the
objects of type `T` and `T2`.
```

**Review by GarveyJoe — APPROVED** [review]

**Review by sergey-semenov — APPROVED** [review]

**Review by broxigarchen — APPROVED** [review]
LGTM!

**Pull request merged**
